From 0eaacbada1b7da1ae4a82ff440aaf7d58e532dd3 Mon Sep 17 00:00:00 2001
From: Phaedrus Leeds <mwleeds@protonmail.com>
Date: Tue, 8 Feb 2022 09:47:10 -0800
Subject: [PATCH] Bring back in-house icon validator

This reverts commit 2fb498f92f8e1c41049dedf887690083fa12f6d1 and makes
related changes so that we use our own copy of Flatpak's icon validator
(taken from icon-validator/validate-icon.c in Flatpak 1.13.1).

Here are the reasons for bringing back this copied code:
(1) To avoid a circular dependency between x-d-p and Flatpak. Debian has
been patching this code back in for years for this purpose.
(2) We don't want icon validation to be optional since it's a security
measure. Having our own binary makes it easier to make it required.
(3) For the new dynamic launcher portal, we need to determine the size
and format of icons. The validator already gathers this information so
it makes sense to just have it print those things to stdout. This means
our copy of this code will diverge from Flatpak.
(4) It is too small to be worth maintaining as a separate repo.
---
 configure.ac                |   1 +
 src/Makefile.am.inc         |   5 +
 src/notification.c          |   6 +-
 src/validate-icon.c         | 249 ++++++++++++++++++++++++++++++++++++
 7 files changed, 264 insertions(+), 6 deletions(-)
 create mode 100644 src/validate-icon.c


diff --git a/configure.ac b/configure.ac
index 2b97f9f..bc38875 100644
--- a/configure.ac
+++ b/configure.ac
@@ -172,6 +172,7 @@ fi
 GLIB_TESTS
 
 PKG_CHECK_MODULES(FUSE, [fuse])
+PKG_CHECK_MODULES(GDK_PIXBUF, [gdk-pixbuf-2.0])
 
 AC_CONFIG_FILES([
 Makefile
diff --git a/src/Makefile.am.inc b/src/Makefile.am.inc
index bcc2a90..29f5a1f 100644
--- a/src/Makefile.am.inc
+++ b/src/Makefile.am.inc
@@ -1,5 +1,6 @@
 libexec_PROGRAMS += \
 	xdg-desktop-portal \
+	xdg-desktop-portal-validate-icon \
 	$(NULL)
 
 service_in_files += \
@@ -216,3 +217,7 @@ xdg_desktop_portal_CPPFLAGS = \
 	-DGETTEXT_PACKAGE=\"$(GETTEXT_PACKAGE)\"        \
 	-DLOCALEDIR=\"$(localedir)\"			\
 	$(NULL)
+
+xdg_desktop_portal_validate_icon_SOURCES = src/validate-icon.c
+xdg_desktop_portal_validate_icon_LDADD = $(GDK_PIXBUF_LIBS)
+xdg_desktop_portal_validate_icon_CFLAGS = $(GDK_PIXBUF_CFLAGS) -DHELPER=\"$(BWRAP)\" -D_GNU_SOURCE=1
diff --git a/src/notification.c b/src/notification.c
index 63c1970..e3c96ec 100644
--- a/src/notification.c
+++ b/src/notification.c
@@ -365,7 +365,7 @@ validate_icon_more (GVariant *v)
   int status;
   g_autofree char *err = NULL;
   g_autoptr(GError) error = NULL;
-  const char *icon_validator = LIBEXECDIR "/flatpak-validate-icon";
+  const char *icon_validator = LIBEXECDIR "/xdg-desktop-portal-validate-icon";
   const char *args[6];
 
   if (G_IS_THEMED_ICON (icon))
@@ -383,8 +383,8 @@ validate_icon_more (GVariant *v)
 
   if (!g_file_test (icon_validator, G_FILE_TEST_EXISTS))
     {
-      g_debug ("Icon validation: %s not found, accepting icon without further validation.", icon_validator);
-      return TRUE;
+      g_debug ("Icon validation: %s not found, rejecting icon by default.", icon_validator);
+      return FALSE;
     }
 
   bytes = g_bytes_icon_get_bytes (G_BYTES_ICON (icon));
diff --git a/src/validate-icon.c b/src/validate-icon.c
new file mode 100644
index 0000000..3de2568
--- /dev/null
+++ b/src/validate-icon.c
@@ -0,0 +1,249 @@
+/*
+ * Copyright © 2018 Red Hat, Inc
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authors:
+ *       Matthias Clasen <mclasen@redhat.com>
+ */
+
+#include <gdk-pixbuf/gdk-pixbuf.h>
+#include <glib/gstdio.h>
+#include <errno.h>
+#include <unistd.h>
+
+static int
+validate_icon (const char *arg_width,
+               const char *arg_height,
+               const char *filename)
+{
+  GdkPixbufFormat *format;
+  int max_width, max_height;
+  int width, height;
+  const char *name;
+  const char *allowed_formats[] = { "png", "jpeg", "svg", NULL };
+  g_autoptr(GdkPixbuf) pixbuf = NULL;
+  g_autoptr(GError) error = NULL;
+
+  format = gdk_pixbuf_get_file_info (filename, &width, &height);
+  if (format == NULL)
+    {
+      g_printerr ("Format not recognized\n");
+      return 1;
+    }
+
+  name = gdk_pixbuf_format_get_name (format);
+  if (!g_strv_contains (allowed_formats, name))
+    {
+      g_printerr ("Format %s not accepted\n", name);
+      return 1;
+    }
+
+  if (!g_str_equal (name, "svg"))
+    {
+      max_width = g_ascii_strtoll (arg_width, NULL, 10);
+      if (max_width < 16 || max_width > 4096)
+        {
+          g_printerr ("Bad width limit: %s\n", arg_width);
+          return 1;
+        }
+
+      max_height = g_ascii_strtoll (arg_height, NULL, 10);
+      if (max_height < 16 || max_height > 4096)
+        {
+          g_printerr ("Bad height limit: %s\n", arg_height);
+          return 1;
+        }
+    }
+  else
+    {
+      /* Sanity check for vector files */
+      max_height = max_width = 4096;
+    }
+
+  if (width > max_width || height > max_height)
+    {
+      g_printerr ("Image too large (%dx%d). Max. size %dx%d\n", width, height, max_width, max_height);
+      return 1;
+    }
+
+  pixbuf = gdk_pixbuf_new_from_file (filename, &error);
+  if (pixbuf == NULL)
+    {
+      g_printerr ("Failed to load image: %s\n", error->message);
+      return 1;
+    }
+
+  return 0;
+}
+
+G_GNUC_NULL_TERMINATED
+static void
+add_args (GPtrArray *argv_array, ...)
+{
+  va_list args;
+  const char *arg;
+
+  va_start (args, argv_array);
+  while ((arg = va_arg (args, const gchar *)))
+    g_ptr_array_add (argv_array, g_strdup (arg));
+  va_end (args);
+}
+
+const char *
+flatpak_get_bwrap (void)
+{
+  const char *e = g_getenv ("FLATPAK_BWRAP");
+
+  if (e != NULL)
+    return e;
+  return HELPER;
+}
+
+
+static gboolean
+path_is_usrmerged (const char *dir)
+{
+  /* does /dir point to /usr/dir? */
+  g_autofree char *target = NULL;
+  GStatBuf stat_buf_src, stat_buf_target;
+
+  if (g_stat (dir, &stat_buf_src) < 0)
+    return FALSE;
+
+  target = g_strdup_printf ("/usr/%s", dir);
+
+  if (g_stat (target, &stat_buf_target) < 0)
+    return FALSE;
+
+  return (stat_buf_src.st_dev == stat_buf_target.st_dev) &&
+         (stat_buf_src.st_ino == stat_buf_target.st_ino);
+}
+
+static int
+rerun_in_sandbox (const char *arg_width,
+                  const char *arg_height,
+                  const char *filename)
+{
+  const char * const usrmerged_dirs[] = { "bin", "lib32", "lib64", "lib", "sbin" };
+  int i;
+  g_autoptr(GPtrArray) args = g_ptr_array_new_with_free_func (g_free);
+  char validate_icon[PATH_MAX + 1];
+  ssize_t symlink_size;
+
+  symlink_size = readlink ("/proc/self/exe", validate_icon, sizeof (validate_icon) - 1);
+  if (symlink_size < 0 || (size_t) symlink_size >= sizeof (validate_icon))
+    {
+      g_printerr ("Error: failed to read /proc/self/exe\n");
+      return 1;
+    }
+
+  validate_icon[symlink_size] = 0;
+
+  add_args (args,
+            flatpak_get_bwrap (),
+            "--unshare-ipc",
+            "--unshare-net",
+            "--unshare-pid",
+            "--ro-bind", "/usr", "/usr",
+            "--ro-bind-try", "/etc/ld.so.cache", "/etc/ld.so.cache",
+            "--ro-bind", validate_icon, validate_icon,
+            NULL);
+
+  /* These directories might be symlinks into /usr/... */
+  for (i = 0; i < G_N_ELEMENTS (usrmerged_dirs); i++)
+    {
+      g_autofree char *absolute_dir = g_strdup_printf ("/%s", usrmerged_dirs[i]);
+
+      if (!g_file_test (absolute_dir, G_FILE_TEST_EXISTS))
+        continue;
+
+      if (path_is_usrmerged (absolute_dir))
+        {
+          g_autofree char *symlink_target = g_strdup_printf ("/usr/%s", absolute_dir);
+
+          add_args (args,
+                    "--symlink", symlink_target, absolute_dir,
+                    NULL);
+        }
+      else
+        {
+          add_args (args,
+                    "--ro-bind", absolute_dir, absolute_dir,
+                    NULL);
+        }
+    }
+
+  add_args (args,
+            "--tmpfs", "/tmp",
+            "--proc", "/proc",
+            "--dev", "/dev",
+            "--chdir", "/",
+            "--setenv", "GIO_USE_VFS", "local",
+            "--unsetenv", "TMPDIR",
+            "--die-with-parent",
+            "--ro-bind", filename, filename,
+            NULL);
+
+  if (g_getenv ("G_MESSAGES_DEBUG"))
+    add_args (args, "--setenv", "G_MESSAGES_DEBUG", g_getenv ("G_MESSAGES_DEBUG"), NULL);
+  if (g_getenv ("G_MESSAGES_PREFIXED"))
+    add_args (args, "--setenv", "G_MESSAGES_PREFIXED", g_getenv ("G_MESSAGES_PREFIXED"), NULL);
+
+  add_args (args, validate_icon, arg_width, arg_height, filename, NULL);
+  g_ptr_array_add (args, NULL);
+
+  {
+    g_autofree char *cmdline = g_strjoinv (" ", (char **) args->pdata);
+    g_debug ("Icon validation: Spawning %s", cmdline);
+  }
+
+  execvpe (flatpak_get_bwrap (), (char **) args->pdata, NULL);
+  /* If we get here, then execvpe() failed. */
+  g_printerr ("Icon validation: execvpe %s: %s\n", flatpak_get_bwrap (), g_strerror (errno));
+  return 1;
+}
+
+static gboolean opt_sandbox;
+
+static GOptionEntry entries[] = {
+  { "sandbox", 0, 0, G_OPTION_ARG_NONE, &opt_sandbox, "Run in a sandbox", NULL },
+  { NULL }
+};
+
+int
+main (int argc, char *argv[])
+{
+  GOptionContext *context;
+  GError *error = NULL;
+
+  context = g_option_context_new ("WIDTH HEIGHT PATH");
+  g_option_context_add_main_entries (context, entries, NULL);
+  if (!g_option_context_parse (context, &argc, &argv, &error))
+    {
+      g_printerr ("Error: %s\n", error->message);
+      return 1;
+    }
+
+  if (argc != 4)
+    {
+      g_printerr ("Usage: %s [OPTION…] WIDTH HEIGHT PATH\n", argv[0]);
+      return 1;
+    }
+
+  if (opt_sandbox)
+    return rerun_in_sandbox (argv[1], argv[2], argv[3]);
+  else
+    return validate_icon (argv[1], argv[2], argv[3]);
+}
